### PR TITLE
Auto Retry (Idempotency requests)

### PR DESCRIPTION
**Pull request by gildub**
Auto Retry option for all services

Auto Retry is off by default but can be turned on with `:idempotent` option.
Default values are `:retry_limit => 2` and `:retry_interval => 5` (seconds) 

Only GET and DELETE REST verbs are used for auto retry. More verbs can be added if that make sense.

Addresses https://github.com/fog/fog-openstack/issues/453

### CODE REVIEW

File: lib/fog/openstack/core.rb [hunk 1]
```diff
@@ -172,6 +180,11 @@ def setup(options)
           instance_variable_set "@#{openstack_param}".to_sym, value
         end
 
+        @idempotent_verbs ||= %w(GET DELETE)
+        @idempotent ||= options[:idempotent].nil? ? false : !!options[:idempotent]
+        @retry_interval ||= options[:retry_interval] ? options[:retry_interval] : 5 # Seconds
+        @retry_limit ||= options[:retry_limit] ? options[:retry_limit] : 2
+
```
Comment (aufi) [inline lib/fog/openstack/core.rb]: Just a nit, but if I'm not mistaken, 3 lines above could be written in shorter way (also matching to pattern few lines bellow)
```ruby
@idempotent     ||= !!options[:idempotent]
@retry_interval ||= options[:retry_interval] || 5 # Seconds
@retry_limit    ||= options[:retry_limit] || 2
```

Comment (gildub) [reply lib/fog/openstack/core.rb]: That's neat.

Absolutely, I'lll looking into adding some tests

Comment (gildub) [reply lib/fog/openstack/core.rb]: @aufi, sorry didn't get the bandwidth to do it.

**Comment by aufi** [comment]
@gildub Hi Gilles, I know it is late, but I'm happy merging this PR when the https://github.com/fog/fog-openstack/pull/462/conflicts gets resolved.

**Comment by gildub** [comment]
@aufi, Hi Markek, sorry for slow response. I've addressed conflict. Thanks!

**Comment by PatrickFranken** [comment]
@aufi any chance to merge this into the master? The Swift service used by my company is pretty unreliable and this should solve that issue.

**Comment by ares** [comment]
@ShamoX this seems as a good and safe addition and ready to get in. Of course tests are missing, what is your preference here?

**Comment by ShamoX** [comment]
Yes, tests are missing. I don't really see how we can tests this except by implementing a complexe request/response scenario, which seems not easy to implement with VCRs (maybe with a sort of interaction with doubles - but seems really complicated).

Anyway, I can review it more precisely when I have more time and see if I can catch any statically analysable bad behaviours. But I doubt it.

We should at least test new configuration variables to see if they are used.

**Pull request closed without merging**